src/userSync.js

```diff
@@ -0,0 +1,171 @@
+      image: [],
+      iframe: []
+    }
+  }
+
+  /**
+   * @function fireSyncs
+   * @summary Trigger all user syncs in the queue
+   * @private
+   */
+  function fireSyncs() {
+    if (!userSyncConfig.syncEnabled || !cookiesAreSupported || hasFired) {
+      return;
+    }
+
+    try {
+      // Image pixels
+      fireImagePixels();
+      // Iframe syncs
+      loadIframes();
+    } catch (e) {
+      return utils.logError('Error firing user syncs', e);
+    }
+    // Reset the user sync queue
+    queue = getDefaultQueue();
+    hasFired = true;
+  }
+
+  /**
+   * @function fireImagePixels
+   * @summary Loops through user sync pixels and fires each one
+   * @private
+   */
+  function fireImagePixels() {
+    if (!userSyncConfig.pixelEnabled) {
+      return;
+    }
+    // Randomize the order of the pixels before firing
+    // This is to avoid giving any bidder who has registered multiple syncs
+    // any preferential treatment and balancing them out
+    utils.shuffle(queue.image).forEach((sync) => {
+      let [bidderName, trackingPixelUrl] = sync;
+      utils.logMessage(`Invoking image pixel user sync for bidder: ${bidderName}`);
+      // Create image object and add the src url
+      utils.triggerPixel(trackingPixelUrl);
+    });
+  }
+
+  /**
+   * @function loadIframes
+   * @summary Loops through iframe syncs and loads an iframe element into the page
+   * @private
+   */
+  function loadIframes() {
+    if (!userSyncConfig.iframeEnabled) {
+      return;
+    }
+    // Randomize the order of these syncs just like the pixels above
+    utils.shuffle(queue.iframe).forEach((sync) => {
+      let [bidderName, iframeUrl] = sync;
+      utils.logMessage(`Invoking iframe user sync for bidder: ${bidderName}`);
+      // Create image object and add the src url
+      utils.insertUserSyncIframe(iframeUrl);
+    });
+  }
+
+  /**
+   * @function incrementAdapterBids
+   * @summary Increment the count of user syncs queue for the adapter
+   * @private
+   * @params {object} numAdapterBids The object contain counts for all adapters
+   * @params {string} bidder The name of the bidder adding a sync
+   * @returns {object} The updated version of numAdapterBids
+   */
+  function incrementAdapterBids(numAdapterBids, bidder) {
+    if (!numAdapterBids[bidder]) {
+      numAdapterBids[bidder] = 1;
+    } else {
+      numAdapterBids[bidder] += 1;
+    }
+    return numAdapterBids;
+  }
+
+  /**
+   * @function registerSync
+   * @summary Add sync for this bidder to a queue to be fired later
+   * @public
+   * @params {string} type The type of the sync including image, iframe
+   * @params {string} bidder The name of the adapter. e.g. "rubicon"
+   * @params {string|object} data A series of arguments
+
+   * @example <caption>Using Image Sync</caption>
+   * // registerSync(type, adapter, pixelUrl)
+   * userSync.registerSync('image', 'rubicon', 'http://example.com/pixel')
+   */
+  publicApi.registerSync = (type, bidder, ...data) => {
```

Comment:
Would strongly prefer a `@param {SyncData} data`, with a `@typedef {object} SyncData` (and explicit params) here.

As someone who wants to call this method, this is not self-documenting. Rules like "the first arg of this `...data` is the URL" are invisible, and the only way to figure them out is to read the code.

Since this method is adapter-facing, I think it's very important to make this as clear as possible.

Reply:
Since we are no longer support AJAX we can make this fix without causing a mess. I will do it now. AJAX support made it complicated.